truth_justice_cake_cutting.py

```diff
@@ -0,0 +1,340 @@
+  def FindPath(self, source, target, path):
+    if source == target:
+      return path
+    for edge in self.GetEdges(source):
+      residual = edge.capacity - self.flow[edge]
+      if residual > 0 and not (edge, residual) in path:
+        result = self.FindPath(edge.target, target, path + [(edge, residual)])
+        if result != None:
+          return result
+
+  def MaxFlow(self, source, target):
+    path = self.FindPath(source, target, [])
+    while path != None:
+      flow = min(res for edge, res in path)
+      for edge, res in path:
+        self.flow[edge] += flow
+        self.flow[edge.redge] -= flow
+      path = self.FindPath(source, target, [])
+    return sum(self.flow[edge] for edge in self.GetEdges(source))
+
+
+
+
+"""
+agents: list of PiecewiseLinearAgent
+"""
+def algorithm2(agents):
+    """
+    >>> a = PiecewiseLinearAgent([(0,0.39, 1,1)], "a")
+    >>> b = PiecewiseLinearAgent([(0,0.6, 1,1)], "b")
+    >>> c = PiecewiseLinearAgent([(0,0.1, 1,1)], "c")
+    >>> agents = [a,b,c]
+    
+    >>> allocations = algorithm2(agents)
+    
+    >>> print(allocations)
+    
+    algorithm2
+    > a gets [(0.0, 0.2), (1.0, 1.2000000000000002), (0.2, 0.3), (0.7, 0.7999999999999998)] with value 2.30
+    > b gets [(0.2, 0.4), (0.8, 1.0), (0.3, 0.39999999999999997), (0.5999999999999999, 0.7)] with value 2.55
+    """
+    print("algorithm2")
+    N = len(agents)
+    logger.info("N is: %d", N)
+    
+    #organise the intervals
+    intervals = []
+    for agent in agents:
+        regions = agent.desired_regions
+        for region in regions:
+            intervals.append((region[0],'start'))
+            intervals.append((region[1],'end'))
+    intervals = sorted(intervals, key = first_element)
+    
+    partitions = [[] for _ in range(N)]
+    startCounter = 0
+    lastPoint = None
+    for i in intervals:
+        if(lastPoint == None):
+            lastPoint = i[0]
+        if(startCounter > 0 and lastPoint != i[0]):
+            for index in range(len(agents)):
+                I = (i[0] - lastPoint) / 2*N
+                
+                partitions[index].append((lastPoint + I*index, lastPoint + (index+1)*I))
+                partitions[index].append((lastPoint + (2*N-index+1)*I, lastPoint + (2*N-index+2)*I))
+
+        if(i[1] == 'start'):
+            startCounter += 1
+        elif(i[1] == 'end'):
+            startCounter -= 1
+        lastPoint = i[0]
+    
+    #random match
+    shuffle(partitions)
+    
+    #print(partitions)
+    
+    results = Allocation(agents)
+    for i in range(N):
+        results.set_piece(i, partitions[i])
+        #results.append(('agent '+str(i), partitions[i]))
+    return results
+
+"""
+agents: list of PiecewiseUniformAgent
+cake: list of intervals
+allocations: should be empty
+
+called by algorithm1
+"""
+def algorithm1_recursive(agents, cake, allocations = {}):
+    # all agents got their cake
+    if not agents:
+        logger.info("Done")
+        return allocations
+    
```

Comment:
Please add more logger.info messages, so it will be easier to  understand how the algorithm works.